### ISSUE TITLE
Deploy infrastructure short-term process change proposal

### ISSUE THREAD

**Issue by bkase** [MinaProtocol/mina#6352]
## Background

Several folks (including myself) are finding ourselves wasting time on simple mistakes when working with the current state of our deployment code. The infrastructure team will be iterating on tooling to mitigate some of these pain points and others, but there are simple process changes we can make to potentially mitigate them immediately. These changes impact the workflows of everyone deploying networks — so rather than just making the changes optimistically, I'm providing this doc where folks can leave comments

## Example Recent Mitigatable Problems

(I've cherry-picked these here that I feel we can at least partially address immediately)

- Testnets failing with random chain-id mismatches and causing mass confusion and debugging effort

On Friday, we realized that this was likely caused because we had started deploying with the docker image tag for branches that did not include a SHA hash. The problem with this: when a container dies, if a new commit had been pushed to the branch (or if even develop had been merged in, for example), the container would be restarted with a separate build. Nathan instantly identified this when he noticed this was how we were configuring our deploys, but this was a week or two after we wasted time looking in the wrong places. 

- Misuse of helm caused wasted QA time

Last week, I attempted to test some libp2p changes that required changing helm charts and terraform. Terraform helpfully fails with validation errors if you use variables incorrectly, but helm silently succeeds. I accidentally used a helm global variable instead of one scoped to a block producer. Because of this, 2 days of QA-net running time were wasted and weren't actually testing what it was supposed to test.

- Checking-in personal edits / working with checked-in personal edits on scripts (like auto-deploy) causing botched deploys

It is common to tweak some of the scripts we use to deploy and monitor the health of nodes. Commonly, these changes that should only be checked-in locally are accidentally pushed. I have both been burned by this and burned others with this. While not as terrible as the other problems, this still causes a good amount of wasted time on failed deploys (given it takes ~10minutes per deploy).

### Turn code reviews on and move branches around

EDIT: Only have master

Firstly, make `pickles-nightly` -> `master`. Master should be what we deploy from, and master only. No deploys allowed from side-branches. Details of how to merge pickles-nightly and master will be fleshed out with a small meeting.

Make `coda-automation` `master` have mandatory code reviews.  This has the potential to catch easy mistakes like the chain-id one or helm misuse. It also can prevent us from checking-in bad changes to scripts. Be sure to tag relevant folks on reviews; moreover, the changes should be small enough that a reviewer should make sure to be thoughtful — this is not useful if the extra pair of eyes doesn't have the right context or is taking care.

In a world where we build new deployments by copy-and-pasting directories, we should make sure to commit the copied directory *before* committing any changes, so code reviews can easily see the real changes, and extra eyes can actually be effective at catching some of these issues.

## Request for more suggested actions + feedback

Turning code reviews on will be a major impact on folks' workflows so I wanted to make sure to provide enough context here for why it is important and open this up to feedback before we enable it. I've personally already lost enough time to stupid mistakes that I feel would vastly cost more than if you were to add up the time waiting on code reviews for months and months. 

Please give feedback! Additionally, please add more suggested workflow changes that can help us in the short term if you can come up with any. For anything that should require more involved fixes, ping Andrew, he has a doc tracking these.

**Comment by nholland94** [comment]
- pre push hook
- sha check tf
- code review policy

Spoke with @bkase some about how we can set some policies here to help us be able to iterate quickly on deployments while still following a review process when necessary. We came up with the following ideas:

1) Any changes which only effect testnet configuration (i.e. `terraform/testnets/**/*`) should be pushed directly to the `master` branch. This push should be done before a deploy is performed, so that others can checkout the same terraform configuration if they need to hack on that deploy.
2) All other changes ("feature changes") should go in separate PRs which are reviewed and merged into `master` before deploys are done with those features.

In order to help enforce #1, we will add a required git hook to the repository which will do a pre-push check to ensure that you are not pushing illegal changes directly to `master`. This will sort of be a "scouts honor" system where we trust people do not add `-n` to their git commands when pushing changes on this branch.

This process leaves holes where configuration issues are not reviewed, so easy configuration mistakes can still be made. In order to alleviate this, we can add more code to terraform which will check for these configuration mistakes and error out if they are made. For example, an easy mistake to make right now is to provide a sha-less docker image name to deploy, which can cause your nodes to redeploy with incompatible images if they are restarted by kubernetes. This kind of mistake can be easily caught in terraform, and there can even be an extra flag which allows you to explicitly opt in to skipping this check if you still want to deploy a sha-less image name.
